File: android/app/src/main/java/com/ethosprotocol/api/VaultWebSocketClient.kt
package com.ethosprotocol.api

// #110 — WebSocket event stream stub.
// Message schema defined in shared/api-contract.md §WebSocket Message Schema.

import com.ethosprotocol.models.Vault
import io.ktor.client.*
import io.ktor.client.engine.android.*
import io.ktor.client.plugins.websocket.*
import io.ktor.serialization.kotlinx.*
import io.ktor.websocket.*
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.isActive
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import javax.inject.Inject
import javax.inject.Singleton

// MARK: - WebSocket event types

sealed class VaultWebSocketEvent {
    /** Server-side vault state changed (TTL refresh, check-in, deposit, etc.). */
    data class VaultUpdated(val vaultId: String, val vault: Vault) : VaultWebSocketEvent()
    /** Vault transitioned to `expired` status. */
    data class VaultExpired(val vaultId: String, val expiredAt: String) : VaultWebSocketEvent()
    /** Funds released to the beneficiary. */
    data class VaultReleased(val vaultId: String, val releasedAt: String, val amount: Long) : VaultWebSocketEvent()
    /** Server keepalive — no action required; client may reply with `pong`. */
    object Ping : VaultWebSocketEvent()
    /** Server signalled a recoverable error. */
    data class Error(val code: String, val message: String) : VaultWebSocketEvent()
    /** Connection closed; will reconnect unless code == 4401. */
    data class Disconnected(val reason: String) : VaultWebSocketEvent()
}

// MARK: - Raw server message envelopes (internal, for decoding only)

@Serializable
private data class WSVaultUpdatedMessage(
    @SerialName("vault_id") val vaultId: String,
    val vault: Vault
)

@Serializable
private data class WSVaultExpiredMessage(
    @SerialName("vault_id") val vaultId: String,
    @SerialName("expired_at") val expiredAt: String
)

@Serializable
private data class WSVaultReleasedMessage(
    @SerialName("vault_id") val vaultId: String,
    @SerialName("released_at") val releasedAt: String,
    val amount: Long
)

@Serializable
private data class WSErrorMessage(
    val code: String,
    val message: String
)

// MARK: - VaultWebSocketClient

/**
 * Stub WebSocket client for real-time vault events (#110).
 *
 * Emits a [Flow] of [VaultWebSocketEvent]s. Reconnects with exponential backoff
 * (1 s base, 60 s cap) on any non-4401 close.
 *
 * NOTE: This is a stub — the connection logic is correct but no staging WebSocket
 * endpoint exists yet. Wire up the Flow into [VaultViewModel] once it does.
 *
 * Usage:
 * ```kotlin
 * wsClient.events(vaultId = "vault-123").collect { event ->
 *     when (event) { ... }
 * }
 * ```
 */
@Singleton
class VaultWebSocketClient @Inject constructor(
    private val tokenProvider: TokenProvider,
    private val baseUrl: String
) {
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    private val client = HttpClient(Android) {
        install(WebSockets) {
            contentConverter = KotlinxWebsocketSerializationConverter(json)
        }
    }

    // Reconnect backoff: doubles each attempt, capped at maxBackoffMs.
    private val baseBackoffMs = 1_000L
    private val maxBackoffMs = 60_000L

    /**
     * Returns a cold [Flow] that connects to the vault WebSocket, emits events,
     * and reconnects automatically on transient failures.
     */
    fun events(vaultId: String): Flow<VaultWebSocketEvent> = callbackFlow {
        var backoffMs = baseBackoffMs
        while (isActive) {
            try {
                val wsUrl = baseUrl
                    .replace("https://", "wss://")
                    .replace("http://", "ws://") + "/ws?vault_id=$vaultId"

                client.webSocket(urlString = wsUrl, request = {
                    tokenProvider.token?.let { headers.append("Authorization", "Bearer $it") }
                }) {
                    backoffMs = baseBackoffMs // reset on successful connection

                    for (frame in incoming) {
                        when (frame) {
                            is Frame.Text -> {
                                val event = parseEvent(frame.readText())
                                if (event != null) send(event)
                                // Reply to keepalive pings
                                if (event == VaultWebSocketEvent.Ping) {
                                    outgoing.trySend(Frame.Text("""{"type":"pong"}"""))
                                }
                            }
                            is Frame.Close -> {
                                val reason = frame.readReason()
                                send(VaultWebSocketEvent.Disconnected(reason?.message ?: "closed"))
                                // Code 4401 == auth failure — do not reconnect
                                if (reason?.code?.toInt() == 4401) {
                                    close()
                                    return@webSocket
                                }
                            }
                            else -> Unit
                        }
                    }
                }
            } catch (e: Exception) {
                send(VaultWebSocketEvent.Disconnected(e.message ?: "error"))
            }

            if (!isActive) break
            delay(backoffMs)
            backoffMs = minOf(backoffMs * 2, maxBackoffMs)
        }
        awaitClose { client.close() }
    }

    // MARK: - Private

    private fun parseEvent(text: String): VaultWebSocketEvent? {
        return try {
            val obj = json.parseToJsonElement(text) as? JsonObject ?: return null
            when (obj["type"]?.jsonPrimitive?.content) {
                "vault_updated" -> {
                    val msg = json.decodeFromString<WSVaultUpdatedMessage>(text)
                    VaultWebSocketEvent.VaultUpdated(vaultId = msg.vaultId, vault = msg.vault)
                }
                "vault_expired" -> {
                    val msg = json.decodeFromString<WSVaultExpiredMessage>(text)
                    VaultWebSocketEvent.VaultExpired(vaultId = msg.vaultId, expiredAt = msg.expiredAt)
                }
                "vault_released" -> {
                    val msg = json.decodeFromString<WSVaultReleasedMessage>(text)
                    VaultWebSocketEvent.VaultReleased(
                        vaultId = msg.vaultId, releasedAt = msg.releasedAt, amount = msg.amount
                    )
                }
                "ping" -> VaultWebSocketEvent.Ping
                "error" -> {
                    val msg = json.decodeFromString<WSErrorMessage>(text)
                    VaultWebSocketEvent.Error(code = msg.code, message = msg.message)
                }
                else -> null // unknown type — ignore forward-compatible
            }
        } catch (_: Exception) {
            null
        }
    }
}

File: ios/EthosProtocol/Sources/Services/VaultWebSocketClient.swift
import Foundation

// MARK: - WebSocket Event Types (#110)
//
// All message types defined in shared/api-contract.md §WebSocket Message Schema.
// The server sends a discriminated JSON union keyed on "type".

enum WebSocketEvent {
    case vaultUpdated(vaultID: String, vault: Vault)
    case vaultExpired(vaultID: String, expiredAt: Date)
    case vaultReleased(vaultID: String, releasedAt: Date, amount: Int64)
    case ping
    case error(code: String, message: String)
    case disconnected(closeCode: URLSessionWebSocketTask.CloseCode)
}

// MARK: - Raw server message envelopes (internal, for decoding only)

private struct WSEnvelope: Decodable {
    let type: String
}

private struct WSVaultUpdatedMessage: Decodable {
    let vaultID: String
    let vault: Vault
    enum CodingKeys: String, CodingKey {
        case vaultID = "vault_id"
        case vault
    }
}

private struct WSVaultExpiredMessage: Decodable {
    let vaultID: String
    let expiredAt: Date
    enum CodingKeys: String, CodingKey {
        case vaultID = "vault_id"
        case expiredAt = "expired_at"
    }
}

private struct WSVaultReleasedMessage: Decodable {
    let vaultID: String
    let releasedAt: Date
    let amount: Int64
    enum CodingKeys: String, CodingKey {
        case vaultID = "vault_id"
        case releasedAt = "released_at"
        case amount
    }
}

private struct WSErrorMessage: Decodable {
    let code: String
    let message: String
}

// MARK: - VaultWebSocketClient

/// Stub WebSocket client for real-time vault events (#110).
///
/// Usage:
/// ```swift
/// let ws = VaultWebSocketClient(vaultID: "vault-123", token: authToken)
/// for await event in ws.events {
///     switch event { ... }
/// }
/// ```
///
/// Reconnect policy: exponential backoff starting at 1 s, capped at 60 s.
/// Authentication failures (server close code 4401) are not retried.
///
/// NOTE: This is a stub — the event loop, backoff, and message parsing are
/// structured correctly but no staging WebSocket endpoint exists yet. Connect
/// once one is available and replace the TODO comments below.
final class VaultWebSocketClient {
    private let vaultID: String
    private let baseURL: URL
    private let session: URLSession
    private let decoder: JSONDecoder

    // Reconnect backoff: doubles each attempt, capped at maxBackoff.
    private let baseBackoff: TimeInterval = 1
    private let maxBackoff: TimeInterval = 60

    init(vaultID: String, baseURL: URL? = nil, session: URLSession = .shared) {
        self.vaultID = vaultID
        let base = baseURL ?? URL(string:
            (Bundle.main.object(forInfoDictionaryKey: "API_BASE_URL") as? String
             ?? "https://api.ethos-protocol.app/v1")
            .replacingOccurrences(of: "https://", with: "wss://")
            .replacingOccurrences(of: "http://", with: "ws://")
        )!
        self.baseURL = base
        self.session = session
        decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        decoder.dateDecodingStrategy = .iso8601
    }

    /// Async stream of WebSocket events. Reconnects automatically unless the
    /// server closes with code 4401 (auth failure).
    var events: AsyncStream<WebSocketEvent> {
        AsyncStream { continuation in
            Task {
                var backoff = baseBackoff
                while !Task.isCancelled {
                    do {
                        try await runSession(continuation: continuation)
                        // Clean close — stop reconnecting.
                        break
                    } catch {
                        // TODO: differentiate auth failure (4401) from transient errors
                        // once the staging endpoint is available. For now, always retry.
                        continuation.yield(.error(
                            code: "reconnect",
                            message: "WebSocket disconnected, retrying in \(Int(backoff))s"
                        ))
                        try await Task.sleep(nanoseconds: UInt64(backoff * 1_000_000_000))
                        backoff = min(backoff * 2, maxBackoff)
                    }
                }
                continuation.finish()
            }
        }
    }

    // MARK: - Private

    private func runSession(continuation: AsyncStream<WebSocketEvent>.Continuation) async throws {
        guard var urlComponents = URLComponents(url: baseURL.appendingPathComponent("ws"),
                                                resolvingAgainstBaseURL: false) else {
            throw URLError(.badURL)
        }
        urlComponents.queryItems = [URLQueryItem(name: "vault_id", value: vaultID)]
        guard let url = urlComponents.url else { throw URLError(.badURL) }

        var request = URLRequest(url: url)
        if let token = KeychainService.shared.loadToken() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let task = session.webSocketTask(with: request)
        task.resume()

        // TODO: send `pong` in response to server `ping` once staging endpoint is live.
        defer { task.cancel(with: .normalClosure, reason: nil) }

        while !Task.isCancelled {
            let message = try await task.receive()
            switch message {
            case .string(let text):
                guard let data = text.data(using: .utf8) else { continue }
                if let event = parseEvent(data) { continuation.yield(event) }
            case .data(let data):
                if let event = parseEvent(data) { continuation.yield(event) }
            @unknown default:
                break
            }
        }
    }

    private func parseEvent(_ data: Data) -> WebSocketEvent? {
        guard let envelope = try? decoder.decode(WSEnvelope.self, from: data) else { return nil }
        switch envelope.type {
        case "vault_updated":
            guard let msg = try? decoder.decode(WSVaultUpdatedMessage.self, from: data) else { return nil }
            return .vaultUpdated(vaultID: msg.vaultID, vault: msg.vault)
        case "vault_expired":
            guard let msg = try? decoder.decode(WSVaultExpiredMessage.self, from: data) else { return nil }
            return .vaultExpired(vaultID: msg.vaultID, expiredAt: msg.expiredAt)
        case "vault_released":
            guard let msg = try? decoder.decode(WSVaultReleasedMessage.self, from: data) else { return nil }
            return .vaultReleased(vaultID: msg.vaultID, releasedAt: msg.releasedAt, amount: msg.amount)
        case "ping":
            return .ping
        case "error":
            guard let msg = try? decoder.decode(WSErrorMessage.self, from: data) else { return nil }
            return .error(code: msg.code, message: msg.message)
        default:
            return nil
        }
    }
}
